core/domain/stats_domain.py

```diff
@@ -1351,3 +1352,115 @@ def validate(self):
             raise utils.ValidationError(
                 'calculation_output is too big to be stored (size: %d): %s' % (
                     sys.getsizeof(output_data), str(output_data)))
+
+
+class LearnerAnswerDetails(object):
+    """Domain object that represents the answer details submitted by the
+    learner.
+    """
+
+    def __init__(
+            self, state_reference, entity_type, interaction_id,
+            learner_answer_info_list, schema_version=(
+                feconf.CURRENT_LEARNER_ANSWERS_DETAILS_SCHEMA_VERSION)):
+        """Constructs a LearnerAnswerDetail domain object.
+
+            Args:
+                state_reference: str. The reference to the state.
+                entity_type: str. The type of entity, eg. exploration,
+                    question etc.
+                interaction_id: str. The ID of the interaction.
+                learner_answer_info_list: list. The list of LearnerAnswerInfo
+                    objects and version specified in this object.
+                schema_version: str. The schema version of this object.
+        """
+        self.state_reference = state_reference
+        self.entity_type = entity_type
+        self.interaction_id = interaction_id
+        self.learner_answer_info_list = learner_answer_info_list
+        self.schema_version = schema_version
+
+    def get_learner_answer_info_dict_list(self):
+        """Returns the learner_answer_info_list stored within this object as
+        a list of LearnerAnswerInfo dicts.
+        """
+        return [learner_answer_info.to_dict()
+                for learner_answer_info in self.learner_answer_info_list]
+
+    def validate(self):
+        """Validates LearnerAnswerDetails domain object."""
+
+        if not isinstance(self.state_reference, basestring):
```

Comment:
There is insufficient validation here. Don't just do type validation, do conceptual validation as well. State references must be formatted in a particular way. Entity types can only be one of two things. Interaction ID must correspond to a valid interaction ID in the codebase. All these need to be validated.

Reply:
Done